### ISSUE TITLE
Keepalive should be automatic if there are physical actions.

### ISSUE THREAD

**Issue by edwardalee** [icyphy/lingua-franca#131]
I believe that keepalive should be automatic if there are physical actions.
It shouldn't have to be explicitly specified as a target property.
Any objections if I make this change for the C target?

**Comment by lhstrh** [comment]
If the event queue is empty, and there are only logical actions, no new events can emerge -- unless we're scheduling logical events asynchronously (which would be an abuse of logical actions). In other words, there would be no point in enforcing `keepalive` in that scenario. I think we should eliminate the `keepalive` target property altogether (not just for the C target) and implement the behavior as you suggest; enforce keep alive if there are physical actions.

**Comment by mew2ub** [comment]
I agree with this (both that physical actions -> keepalive and no physical actions -> no keepalive).

I think we might want to mimic the way Node.js handles a similar situation. A Node.js program will stay alive as long as there are event handlers waiting for events, but once the script has finished executing and there are no more event handlers it will exit.

For us, physical actions are like event handlers in Node. To safely get rid of the `keepalive` property we'll need to track physical actions across mutations. It wouldn't be the end of the world if a reactor that started out with physical actions kept running after all physical actions were removed by mutations, but it would definitely be a problem if a reactor with physical actions later added by mutations was prematurely stopped.

**Comment by edwardalee** [comment]
I have realized this issue is more subtle.  When we have physical connections, perhaps notated with ~>, then a physical action will be used, but we may still want keepalive to be false. The physical action will be scheduled synchronously.

I have implemented a mechanism that forces keeplive == true for a federated execution because in this case there will be asynchronously invoked physical actions. But even here, the situation is a bit more subtle. In a federated execution, we probably don't want the *only* mechanism for stopping an execution to be a timeout. So we probably need that a call to stop() should broadcast the request to all federates to request that they stop() at a specified logical time.
